Add connection pooling

jdr0887 opened the pull request.

jdr0887 commented:
@cbizon Mind reviewing this?

maximusunc reviewed (changes requested):
Hi Jason, it looks like a few things were lost when merging the main branch back into your branch. Could you please double check this and make sure you're not overwriting anything from the last merge request please?

maximusunc commented:
A few other comments/questions:
- Just for my understanding, why do we need multiple rabbitmq connections/channels?
- Could you actually delete the `service_aggregator.py` file in the root of the repo please, since we're cleaning up the code? It's confusing to have two of them. (unless that's intentional @cbizon)

jdr0887 commented:
I did the following:

- deleted service_aggregator.py.
- I modified a few merge discrepancies I found

Like database connections, message queuing connections are expensive to create.  Using a connection pool is more resilient & preferable.  This [page](https://www.rabbitmq.com/connections.html) details the lifecycle of a connection.  And this [page](https://www.cloudamqp.com/blog/the-relationship-between-connections-and-channels-in-rabbitmq.html) provides a good explanation between connections & channels.

maximusunc reviewed (changes requested):
This all looks good, just a few minor merging inconsistencies still.

jdr0887 commented on src/service_aggregator.py:
```diff
@@ -306,17 +312,17 @@ async def check_for_messages(guid, pydantic_kgraph, accumulated_results, num_que
                         jr = process_message(message)
                         if is_end_message(jr):
-                            logger.debug(f'{guid}: Received complete message from multistrider')
+                            logger.debug("Received complete message from multistrider")
```
I removed the guid from the logging message since it is redundant with the logging message just above it (line 315).  Maybe switch which line has the guid?

maximusunc replied:
I think it's still very useful. The log on 315 doesn't tell us what type of response we got back. If Strider happens to send back a completed message before we've received all `num_queries`, we would have no idea, and that would be very bad. In general, it's nice to look through the logs and match every log with a uid.

cbizon replied:
Yeah, the log is only comprehensible when every message has its guid - otherwise they intermix and it's just too hard to understand who did what

jdr0887 replied:
I am not sure I agree, but I am not going to quibble over it.  I will add it back.

jdr0887 commented on src/aragorn_app.py:
```diff
@@ -102,56 +134,42 @@ async def subservice_callback(response: PDResponse,  guid: str) -> int:
-
-            # only store file if there's a queue waiting for it
-            # will throw error if queue doesn't exist
-            await channel.get_queue(guid, ensure=True)
```
Is the right way to ensure a queue exists?  Shouldn't it be something like: 

                await channel.declare_queue(guid, durable=False, passive=True, auto_delete=False)

jdr0887 replied:
After reading the documentation for both methods, I think both are valid.  I will add that line back in.

maximusunc replied:
We don't want to create a new queue here, we want to make sure that one already exists before we write a new file to disk. If the queue doesn't exist, that means the original worker that's waiting for this response is no longer waiting for it and no one is going to read and then delete this file, so it's basically a storage space leak. What I was trying to do here was to error out if the queue didn't exist already.

jdr0887 replied:
From the documentation:

async get_queue(name: str, *, ensure: bool = True) → AbstractQueue[[source]](https://aio-pika.readthedocs.io/en/latest/_modules/aio_pika/channel.html#Channel.get_queue)

With ensure=True, it’s a shortcut for .declare_queue(..., passive=True); otherwise, it returns a queue instance without checking its existence.

maximusunc replied:
Their documentation is quite bad. If you read the very next sentence under that one: When the queue does not exist, if ``ensure=True``, will raise :class:`aio_pika.exceptions.ChannelClosed`.
